diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/search/VariablesSearch.kt

```diff
@@ -0,0 +1,95 @@
+package org.cqfn.diktat.ruleset.utils.search
+
+import com.pinterest.ktlint.core.ast.ElementType
+import org.cqfn.diktat.ruleset.utils.*
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getChildrenOfType
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.parents
+import org.jetbrains.kotlin.psi.psiUtil.referenceExpression
+
+/**
+ *  @param node - root node of a type File that is used to search all declared properties (variables)
+ *  it should be ONLY node of File elementType
+ *  @param filterForVariables - condition to filter
+ */
+abstract class VariableSearch(val node: ASTNode, private val filterForVariables: (KtProperty) -> Boolean) {
+
+    /**
+     * to complete implementation of a search mechanism you need to specify what and how you will search in current scope
+     * [this] - scope where to search the usages/assignments/e.t.c of the variable (can be of types KtBlockExpression/KtFile/KtClassBody)
+     */
+    protected abstract fun KtElement.getAllSearchResults(property: KtProperty): List<KtNameReferenceExpression>
+
+    /**
+     * method collects all declared variables and it's usages
+     *
+     * @return a map of a property to it's usages
+     */
+    fun collectVariables(): Map<KtProperty, List<KtNameReferenceExpression>> {
+        require(node.elementType == ElementType.FILE) {
+            "To collect all variables in a file you need to provide file root node"
+        }
+        return node
+                .findAllNodesWithSpecificType(ElementType.PROPERTY)
+                .map { it.psi as KtProperty }
+                .filter(filterForVariables)
+                .associateWith { it.getSearchResults() }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    fun KtProperty.getSearchResults(): List<KtNameReferenceExpression> {
+        return this
+                .getDeclarationScope()
+                // if declaration scope is not null - then we have found out the block where this variable is stored
+                // else - it is a global variable on a file level or a property on the class level
+                .let { declarationScope ->
+                    // searching in the scope with declaration (in the context)
+                    declarationScope?.getAllSearchResults(this)
+                    // searching on the class level in class body
+                            ?: (this.getParentOfType<KtClassBody>(true)?.getAllSearchResults(this))
+                            // searching on the file level
+                            ?: (this.getParentOfType<KtFile>(true)!!.getAllSearchResults(this))
+                }
+    }
+
+    /**
+     * filtering object's fields (expressions) that have same name as variable
+     */
+    protected fun isReferenceToFieldOfObject(expression: KtNameReferenceExpression) =
```

Comment:
Maybe make it an extension on `KtNameReferenceExpression`?

Reply:
fixed